sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt

```diff
@@ -178,13 +195,102 @@ internal class ViewAnnotationManagerImpl(
+        if (north == null || calculateCoordinateBoundForAnnotation(north!!.first, zoom).north() < annotationBounds.north()) {
+          north = Pair(options, frame)
+          isCorrectBound = false
+        }
+        if (east == null || calculateCoordinateBoundForAnnotation(east!!.first, zoom).east() < annotationBounds.east()) {
+          east = Pair(options, frame)
+          isCorrectBound = false
+        }
+        if (south == null || calculateCoordinateBoundForAnnotation(south!!.first, zoom).south() > annotationBounds.south()) {
+          south = Pair(options, frame)
+          isCorrectBound = false
+        }
+        if (west == null || calculateCoordinateBoundForAnnotation(west!!.first, zoom).west() > annotationBounds.west()) {
+          west = Pair(options, frame)
+          isCorrectBound = false
+        }
+      }
+
+      // adding extra checks for nullability (this shouldn't execute normally as we are checking nullability in loop of viewannotations).
+      if (north == null || east == null || south == null || west == null) {
+        logW(TAG, "ViewAnnotation options framing is null. Returning null camera option")
+        return null
+      }
+
+      val coordinateBoundsForCamera = CoordinateBounds(
+        Point.fromLngLat(west!!.coordinate().longitude(), south!!.coordinate().latitude()),
+        Point.fromLngLat(east!!.coordinate().longitude(), north!!.coordinate().latitude())
+      )
+
+      val paddings = EdgeInsets(
+        (edgeInsets?.top ?: 0).toDouble().plus(abs((north!!.second?.top ?: 0.0).toDouble())),
+        (edgeInsets?.left ?: 0).toDouble().plus(abs((west!!.second?.left ?: 0.0).toDouble())),
+        (edgeInsets?.bottom ?: 0).toDouble().plus(abs((south!!.second?.bottom ?: 0.0).toDouble())),
+        (edgeInsets?.right ?: 0).toDouble().plus(abs((east!!.second?.right ?: 0.0).toDouble()))
+      )
+
+      cameraOptionForCoordinates = mapboxMap.cameraForCoordinateBounds(
+        coordinateBoundsForCamera,
+        paddings,
+        bearing,
+        pitch
+      )
+    }
+    return cameraOptionForCoordinates
+  }
+
+  private fun Pair<ViewAnnotationOptions, Rect?>.coordinate() = (first.geometry as Point)
+
+  /**
+   * Function to calculate coordinatebound associated with annotation option.
+   * this uses [MapboxMap.projectedMetersForCoordinate] and [MapboxMap.coordinateForProjectedMeters]
+   * function to get correct coordinate for northEast and southWest point of annotations.
+   */
+  @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+  internal fun calculateCoordinateBoundForAnnotation(
+    viewAnnotationOptions: ViewAnnotationOptions,
+    zoom: Double?
+  ): CoordinateBounds {
+    val frame = getViewAnnotationOptionsFrame(viewAnnotationOptions) ?: Rect(0, 0, 0, 0)
+    val metersPerPixelAtLatitude =
+      if (zoom == null) mapboxMap.getMetersPerPixelAtLatitude((viewAnnotationOptions.geometry as Point).latitude()) else
```

Comment:
Can it actually be null? What if it's called with the `null` value - will it use current zoom level of the map hence get incorrect `MetersPerPixelAtLatitude` value?

Reply:
in our implementation, it can't be `null` (we have checks before we calculate cameraOptions). but this function can be used independently to get the `coordinateBounds` by viewAnnotation. If the zoom is not provided it will use current map zoom to get the coordinatebound. So it's upto the calling function how they want to use it.